### ISSUE TITLE
New dataset

### ISSUE THREAD

**Issue by silvia1993** [wvangansbeke/Unsupervised-Classification#26]
Hi Wouter!
First of all congratulations for your work, it’s really interesting :-) 

I’m trying to use your algorithm with a new dataset: it has ∼4000 images, ∼60 classes and the size of the images is not fixed, I use “random_resized_crop” to 224 for the whole dataset (the size of the images is always greater than 224).

1 - With 224 as image size I have to decrement the batch size (from 512 to 32) since I have memory problems. Your advice is to decrement the image size in such a way that the model is trainable with batch size 512 (or at least 256) or it’s better to have image size 224 and batch size 32? (Referring to page 6 of your paper in the “Implementation details” section I read: “..we approximate the dataset statistics by sampling batches of sufficiently large size.”) 

2 - In the case in which it’s better to have image size 224 and batch size 32, should I change something else? (for example learning rate.. ecc)

3 - I notice that you use a different implementation of Resnet-18 for Cifar and Stl (max pooling and average pooling). It is just a change due to dimensionality problems or there is a different reason? Should I create a custom Resnet-18 also for my new dataset?

4 - Why did you use Moco for ImageNet and SimCLR for Cifar/Stl?

Thank you for your time.

**Comment by wvangansbeke** [comment]
Hi @silvia1993,

Thank you for the kind words and for checking out the code!

It is well known that instance discrimination can learn better representations if more data is available. You have around 67 images per class. This is quite low. CIFAR10 contains 5k images per class, ImageNet 1.3k and STL10 500 for example. You can always start from a network which is pretrained on ImageNet but it is hard to say if this is enough (probably not). I'm also not sure how much your data differs from ImageNet (domain gap).  To learn good representations from a low amount of data is far from solved since this is very difficult. However, I will try to answer your questions here:

1) Using the same augmentations as we did for ImageNet should be fine. A batch size of 32 is likely too low though. I would really advise to use a larger batch size (512 should be fine for 60 classes). It has been shown that instance discrimination performs better with larger batch sizes (i.e. more negatives in the loss) and a larger batch size is also beneficial for SCAN. You can try using Apex (floating point 16), smaller input resolution, smaller network, etc.

2) In general you should scale the learning rate linearly with the batch size, such that a smaller batch size corresponds with a smaller learning rate. This is only valid for sgd though, adam has different rules.

3) Indeed, ReseNet18 is slightly different from the original implementation. We used the same implementation as Wu et al. The input resolution of CIFAR10 is only 32x32 after all.

4) We use the instance discrimination implementation from MoCo because you can train it with a smaller batch size (they adopt a memory bank), which is very useful for ImageNet. Training with batch sizes of 8k on ImageNet is not feasible with the infrastructure we have. SimCLR is fine for smaller datasets with a lower amount of classes though. MoCo is also written in Pytorch which allows us to simply load their pretrained model.  

Hope this helps.

**Comment by silvia1993** [comment]
Thank you for your quick answer, I'll try to take your advice!

**Comment by wvangansbeke** [comment]
Feel free to ask more questions if something is unclear. I'm closing this for now though.

**Comment by silvia1993** [comment]
Hi @wvangansbeke,

I have another small question. I’m trying to fine-tune the network with the SimCLR task starting from the Supervised Pre-trained on Imagenet and it seems that the clustering performance decreases with this type of fine-tuning as if the initial features obtained with the Supervised Imagenet gets ruined. Have you tried this type of training? Do you have an intuition about this behavior? 

Thank you!

**Comment by wvangansbeke** [comment]
Hi @silvia1993,

I never tried clustering starting from the ImageNet pretrained weights because I wanted to keep it completely unsupervised. I don't know if running SimCLR, starting from ImageNet pretrained weights, will help. It seems that performing SimCLR on your dataset hurts. I guess that can happen, but I never encountered something like that. SimCLR is specifically useful when you have large amounts of unlabeled data. I assume that using ImageNet pretrained weights would not hurt the performance in that case. However, it is hard to say because it all depends on the amount and type of data you're using. It might be that you don't have enough data to perform instance discrimination properly. In that case it is not surprising that fine-tuning from pretrained ImageNet weights performs better.
